pyamg/amg_core/linalg.h

```diff
@@ -1103,4 +1125,332 @@ void filter_matrix_rows(const I n_row,
+    for (I i=m; i<n; i++) {
+        x[i] = 0;
+    }
+}
+
+/* Forward substitution solve on lower-triangular linear system,
+ * Lx = rhs, where L is stored in column- or row-major form. 
+ * 
+ * Parameters
+ * ----------
+ *        L : double array, length m*n
+ *            Lower-triangular array stored in column- or row-major.
+ *        rhs : double array, length m
+ *            Right hand side of linear system
+ *        x : double array, length n
+ *            Preallocated array for solution
+ *        m : &int
+ *            Number of rows in L
+ *        n : &int
+ *            Number of columns in L
+ *        is_col_major : bool
+ *            True if L is stored in column-major, false
+ *            if L is stored in row-major.
+ *
+ * Returns
+ * -------
+ *        Nothing, solution is stored in x[].
+ *
+ * Notes
+ * -----
+ * L need not be square, the system will be solved over the
+ * rank r lower-triangular block. If remaining entries in
+ * solution are unused, they will be set to zero. If a zero
+ * is encountered on the ith diagonal, x[i] is set to zero. 
+ *
+ *
+ */
+template<class I, class T>
+void lower_tri_solve(const T L[],
+                     const T rhs[],
+                     T x[],
+                     const I &m,
+                     const I &n,
+                     const I is_col_major)
+{
+    // Function pointer for row or column major matrices
+    I (*get_ind)(const I, const I, const I);
+    const I *C;
+    if (is_col_major) {
+        get_ind = &col_major;
+        C = &m;
+    }
+    else {
+        get_ind = &row_major;
+        C = &n;
+    }
+
+    // Forward substitution
+    I rank = std::min(m,n);
+    for (I i=0; i<rank; i++) {
+        T temp = rhs[i];
+        for (I j=0; j<i; j++) {
+            temp -= L[get_ind(i,j,*C)]*x[j];
+        }
+        if (std::abs(L[get_ind(i,i,*C)]) < 1e-12) {
+            x[i] = 0.0;
+        }
+        else{
+            x[i] = temp / L[get_ind(i,i,*C)];
+        }
+    }
+
+    // If rank < size of rhs, set free elements in x to zero
+    for (I i=m; i<n; i++) {
+        x[i] = 0;
+    }
+}
+
+/* Method to solve the linear least squares problem.
+ *
+ * Parameters
+ * ----------
+ *        A : double array, length m*n
+ *            2d array stored in column- or row-major.
+ *        b : double array, length m
+ *            Right hand side of unconstrained problem.
+ *        x : double array, length n
+ *            Container for solution
+ *        m : &int
+ *            Number of rows in A
+ *        n : &int
+ *            Number of columns in A
+ *        is_col_major : bool
+ *            True if A is stored in column-major, false
+ *            if A is stored in row-major.
+ *
+ * Returns
+ * -------
+ *         x : vector<double>
+ *            Solution to constrained least sqaures problem.
+ *
+ * Notes
+ * -----
+ * If system is under determined, free entries are set to zero. 
+ * Currently only set up for real-valued matrices. May easily
```

Comment:
Should we verify behavior for complex?  Or put a warning in air solver that complex is not supported?

Reply:
that's a good question; I'm inclined to just warn not currently implemented for complex. I dont think there would be any issue, but have never had a complex application that air would make sense for. actually an interesting idea worth thinking about (Ahsan?). In any case, we would need to check the QR and some other inner pieces are correct/robust for complex, so I would punt for now

Reply:
Good idea regarding Ahsan.  I'll put a warning in for now